tracker/src/lock.rs

```diff
@@ -0,0 +1,256 @@
+use lock_api::RawRwLock;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+/// An instrumented Read-Write Lock
+pub type RwLock<T> = lock_api::RwLock<InstrumentRawRwLock<parking_lot::RawRwLock>, T>;
+
+/// A Lock tracker can be used to create instrumented read-write locks
+/// that will record contention metrics
+#[derive(Default, Debug, Clone)]
+pub struct LockTracker {
+    inner: Arc<LockTrackerShared>,
+}
+
+impl LockTracker {
+    pub fn new_lock<T>(&self, t: T) -> RwLock<T> {
+        RwLock::const_new(
+            InstrumentRawRwLock {
+                inner: parking_lot::RawRwLock::INIT,
+                shared: Some(Arc::clone(&self.inner)),
+            },
+            t,
+        )
+    }
+
+    pub fn exclusive_count(&self) -> u64 {
+        self.inner.exclusive_count.load(Ordering::Relaxed)
+    }
+
+    pub fn shared_count(&self) -> u64 {
+        self.inner.shared_count.load(Ordering::Relaxed)
+    }
+
+    pub fn exclusive_wait_nanos(&self) -> u64 {
+        self.inner.exclusive_wait_nanos.load(Ordering::Relaxed)
+    }
+
+    pub fn shared_wait_nanos(&self) -> u64 {
+        self.inner.shared_wait_nanos.load(Ordering::Relaxed)
+    }
+}
+
+#[derive(Debug, Default)]
+struct LockTrackerShared {
+    exclusive_count: AtomicU64,
+    shared_count: AtomicU64,
+    exclusive_wait_nanos: AtomicU64,
+    shared_wait_nanos: AtomicU64,
+}
+
+/// The RAII-goop for locks is provided by lock_api with individual crates
+/// such as parking_lot providing raw lock implementations
+///
+/// This is a raw lock implementation that wraps another and instruments it
+#[derive(Debug)]
+pub struct InstrumentRawRwLock<R: Sized> {
+    inner: R,
+    shared: Option<Arc<LockTrackerShared>>,
+}
+
+/// # Safety
+///
+/// Implementations of this trait must ensure that the `RwLock` is actually
+/// exclusive: an exclusive lock can't be acquired while an exclusive or shared
+/// lock exists, and a shared lock can't be acquire while an exclusive lock
+/// exists.
+///
+/// This is done by delegating to the wrapped RawRwLock implementation
+unsafe impl<R: RawRwLock + Sized> lock_api::RawRwLock for InstrumentRawRwLock<R> {
+    const INIT: Self = Self {
+        inner: R::INIT,
+        shared: None,
+    };
+    type GuardMarker = R::GuardMarker;
+
+    /// Acquires a shared lock, blocking the current thread until it is able to do so.
+    fn lock_shared(&self) {
+        match &self.shared {
+            Some(shared) => {
+                // Early return if possible - Instant::now is not necessarily cheap
+                if self.try_lock_shared() {
```

Comment:
Did you intend to instrument this internal check? Including calls to `lock_shared()` in the `try_lock_shared()` metrics will mean we'll have to do the maths to figure out how many calls the application makes to `try_lock_shared`.

```suggestion
                if self.inner.try_lock_shared() {
```

Reply:
The metric is the number of times a shared lock is acquired by whatever means, so I think this is correct? If it is able to acquire a lock it will increment the shared_count in try_lock_shared and then return from here?

Reply:
I suppose it depends what you're trying to measure - I thought we were measuring "the number of calls the application made to lock methods" but I think you're saying you want to record "number of times the lock methods are invoked, including internal overhead calls made by the instrumentation wrapper"?

Reply:
I want to measure number of times the lock is acquired by any means :grin:

Reply:
Is it not better represented as a call to `lock_shared` with a duration of `0ns`? The application calls `lock_shared` and didn't have to wait.

Reply:
That's what this does? No?

Reply:
Thank you for your helpful comment - I thought `try_lock_shared` was updating it's own counter - in that case, do we not care about recording `try_lock_shared` separately then? I suppose it's not an important distinction unless someone has rolled their own terrible spinlock.

Reply:
Yeah if people are rolling their own spinlock using try_lock_shared that's their problem imo :grin: